### ISSUE TITLE
Performance issues after 2.14

### ISSUE THREAD

**Issue by GCastilho** [prisma/prisma#5792]
## Bug description

After upgrading prisma do a version higher than 2.14 I'm experiencing performance issues with the queries. Mocha reports tests that usually took less than 60ms to now be taking more than 150ms, sometimes more than 200ms to complete

I commented about this in #5485 and analyzed the logs, but they're identical (or very similar) so it does not seems like it's the same as #5274

## Expected behavior

Average performance should remain the same or improve between Prisma versions.

## Prisma information

I've picked some queries to show comparing 2.14 with 2.15 and 2.17, but they were _exactly_ the same, the only difference was the time it took to run them (54ms -> 141ms, 65ms -> 193ms) so I don't thing it make sense to show them (since they're the same). But it's not a problem to share that part of the schema and queries if needed

## Environment & setup
- OS: Debian 10 buster x64
- Database: PostgreSQL 12 (localhost)
- Node.js version: v14.15.5
- Prisma version: >= 2.15.0

Please let me know if you need more information or tests that you need. Thanks!

**Comment by timsuchanek** [comment]
Thanks for reporting @GCastilho!
Could you share the specific query with us plus a subset of the schema that got slower?

**Comment by GCastilho** [comment]
For sure, @timsuchanek 

This is the schema:

```prisma
model User {
  @@map("users")
  id              String   @id @default(uuid())
  email           String   @unique
  password        String
  role            UserRole @default(USER)
  created_at      DateTime @default(now())
  updated_at      DateTime @updatedAt
  lotes           Lote[]
  criadores       Criador[]
  granjas         Granja[]
}

model Criador {
  @@map("criadores")
  id          Int
  userId      String
  user        User      @relation(fields: [userId])
  nome        String
  telefone    String
  created_at  DateTime  @default(now())
  updated_at  DateTime  @updatedAt
  granjas     Granja[]
  lotes       Lote[]
  @@id([userId, id])
}
```

And the queries logged by prisma:

From the first test, prisma@2.15:

```sql
prisma:info  Starting a postgresql pool with 5 connections.
prisma:info  Fetched a connection from the pool
prisma:info  Started http server on http://127.0.0.1:34325
prisma:info  Fetched a connection from the pool
prisma:query DELETE FROM users
prisma:info  Fetched a connection from the pool
prisma:query BEGIN
prisma:query INSERT INTO "public"."users" ("id","email","password","role","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6) RETURNING "public"."users"."id"
prisma:query SELECT "public"."users"."id", "public"."users"."email", "public"."users"."password", "public"."users"."role", "public"."users"."created_at", "public"."users"."updated_at" FROM "public"."users" WHERE "public"."users"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
prisma:info  Fetched a connection from the pool
prisma:query SELECT "public"."users"."id", "public"."users"."password", "public"."users"."role" FROM "public"."users" WHERE "public"."users"."email" = $1 LIMIT $2 OFFSET $3
prisma:info  Fetched a connection from the pool
prisma:query DELETE FROM criadores
prisma:info  Fetched a connection from the pool
prisma:query BEGIN
prisma:query SELECT "public"."users"."id" FROM "public"."users" WHERE "public"."users"."id" = $1 OFFSET $2
prisma:query INSERT INTO "public"."criadores" ("id","userId","nome","telefone","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6) RETURNING "public"."criadores"."userId", "public"."criadores"."id"
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id", "public"."criadores"."nome", "public"."criadores"."telefone", "public"."criadores"."created_at", "public"."criadores"."updated_at" FROM "public"."criadores" WHERE ("public"."criadores"."userId" = $1 AND "public"."criadores"."id" = $2) LIMIT $3 OFFSET $4
prisma:query COMMIT
prisma:info  Fetched a connection from the pool
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id", "public"."criadores"."nome", "public"."criadores"."telefone", "public"."criadores"."created_at", "public"."criadores"."updated_at" FROM "public"."criadores" WHERE "public"."criadores"."id" = $1 LIMIT $2 OFFSET $3
```

From the second test:

```sql
prisma:info  Fetched a connection from the pool
prisma:query DELETE FROM criadores
prisma:info  Fetched a connection from the pool
prisma:query BEGIN
prisma:query SELECT "public"."users"."id" FROM "public"."users" WHERE "public"."users"."id" = $1 OFFSET $2
prisma:query INSERT INTO "public"."criadores" ("id","userId","nome","telefone","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6) RETURNING "public"."criadores"."userId", "public"."criadores"."id"
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id", "public"."criadores"."nome", "public"."criadores"."telefone", "public"."criadores"."created_at", "public"."criadores"."updated_at" FROM "public"."criadores" WHERE ("public"."criadores"."userId" = $1 AND "public"."criadores"."id" = $2) LIMIT $3 OFFSET $4
prisma:query COMMIT
prisma:info  Fetched a connection from the pool
prisma:query BEGIN
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id" FROM "public"."criadores" WHERE ("public"."criadores"."userId" = $1 AND "public"."criadores"."id" = $2)
prisma:query UPDATE "public"."criadores" SET "nome" = $1, "updated_at" = $2 WHERE ("public"."criadores"."userId","public"."criadores"."id") IN (($3,$4))
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id", "public"."criadores"."nome", "public"."criadores"."telefone", "public"."criadores"."created_at", "public"."criadores"."updated_at" FROM "public"."criadores" WHERE ("public"."criadores"."userId" = $1 AND "public"."criadores"."id" = $2) LIMIT $3 OFFSET $4
prisma:query COMMIT
prisma:info  Fetched a connection from the pool
prisma:query SELECT "public"."criadores"."userId", "public"."criadores"."id", "public"."criadores"."nome", "public"."criadores"."telefone", "public"."criadores"."created_at", "public"."criadores"."updated_at" FROM "public"."criadores" WHERE "public"."criadores"."id" = $1 LIMIT $2 OFFSET $3
```

The tests were performed using mocha and the generated "prisma:query" are identical between 2.14, 2.15 and 2.17, the only difference (just noticed it) are the "prisma:info" statements, where on 2.14 we only had these two:

```
prisma:info  Starting a postgresql pool with 5 connections.
prisma:info  Started http server on http+unix:///tmp/prisma-b055a1ccfdb2a3a62e187d82.sock
```

**Comment by timsuchanek** [comment]
Thanks @GCastilho!
One more thing: Could you show the Prisma Client query that you are sending?
So in the form like this:
```ts
prisma.user.findMany({ where: {} })
```

Thanks!

**Comment by GCastilho** [comment]
For sure, @timsuchanek 

Here are the queries, in order:

```typescript
await prisma.user.create({
	data: {
		email,
		password,
		role,
	}
})

await prisma.user.findUnique({
	select: { id: true, password: true, role: true },
	where: { email },
})

await prisma.criador.create({
	data: {
		...novoCriador,
		userId,
	}
})

await prisma.criador.findFirst({
	where: { id: 1 }
})

await prisma.criador.update({
	where: {
		userId_id: {
			userId: req.user.sub,
			id: idCriador
		}
	},
	data: updateCriador
})
```

Duplicated queries I've displayed only once. The delete queries are all used as 

```typescript
await prisma.$queryRaw('DELETE FROM table')
```
Because prisma (AFAIK) still does not have support for "deleteMany" with cascade relations

**Comment by Akxe** [comment]
This does happen to me too! I have shared all info with @matthewmueller via email. My query however did go from less than a second to 25-35sec!

Also, I would like to note, I did test every single query sent to DB and every one finished under 100ms (5 total)

**Comment by Tsvetomir-M** [comment]
A simple query on 2.14 is much faster than on 2.19.
I have tested with this simple example.

Environment:
```
OS: Debian GNU/Linux 9 (stretch) 64-bit
Database: PostgreSQL 13.0 (Debian 13.0-1.pgdg90+1)
Node.js: v14.2.0
```

Db model:
```
model test {
  id   Int    @id @default(autoincrement()) @db.Integer
  name String @db.Text
}
```

Code:
```
let start = new Date().getTime();

for (let i = 0; i < 1000; i++) {
    let res = await prisma.test.findUnique({
        select: {
            id: true,
            name: true,
        },
        where: {
            id: 1
        }
    });
}

console.log(new Date().getTime() - start);
```

Results:
```
Prisma 2.14.0 - 4 seconds
Prisma 2.19.0 - 56 seconds
```

**Comment by chinanderm** [comment]
I've also noticed considerably slower performance with queries.

**Comment by pantharshit00** [comment]
Hello everyone,

Can you all still reproduce these regressions in the latest version of Prisma?

**Comment by Akxe** [comment]
@pantharshit00 Yes, indeed. 

I am doing and measuring 5 simultaneous queries (+ one initial query)

```
[GQL/clientOffer] 📑 "offerInfo": 135.357ms
[GQL/clientOffer] 📑 "carriersPromise": 38.925s
[GQL/clientOffer] 📑 "groups": 4.69ms
[GQL/clientOffer] 📑 "districts": 5.264ms
[GQL/clientOffer] 📑 "motives": 6.566ms
[GQL/clientOffer] 📑 "regions": 6.359ms
[GQL/clientOffer] 📑 "all": 38.933s
```

carriersPromise query:
```ts
{
  select: {
    carrierID: true,
    shortID: true,
    type: true,
    sides: true,
    periods: {
      select: {
        selected: true,
        periodID: true,
        wanted: true,
        reservation: true,
        motiveID: true,
        sellingPrice: true,
        count: true,
        groups: { select: { groupID: true, motiveID: true } }
      }
    },
    carrier: {
      select: {
        lat: true,
        lng: true,
        city: true,
        street: true,
        traffic: true,
        description: true,
        districtID: true,
        regionID: true,
        orientation: true,
        photos: { select: { src: true, order: true } }
      }
    },
    offerID: true
  },
  where: {
    offerID: 9042,
    carrier: { districtID: { in: undefined }, regionID: { in: undefined } }  }
}
```

**Comment by Sytten** [comment]
Honestly the best way to help would be to post the resulting SQL queries for both and verify if there is not a missing index. Hard to help you otherwise.

**Comment by Akxe** [comment]
@Sytten Prisma dev team has full DB with schema and data to test it on. We were in contact previously. All relations are on foreign keys that will automatically create indexes and condition are on primary and normal indexies.

**Comment by GCastilho** [comment]
@pantharshit00 Yes, it's still happening in v2.22.1

**Comment by pantharshit00** [comment]
I used reproduction provided by @Tsvetomir-M to test this out using the following script:
```ts
import { PrismaClient } from "./prisma/client1";
import { PrismaClient as NAPIPrismaClient } from "./prisma/client2";
import { PrismaClient as OLDPrismaClient } from "./prisma/client3";
import { performance } from "perf_hooks";

const prisma = new PrismaClient();
const napi = new NAPIPrismaClient();
const old = new OLDPrismaClient();
async function main() {
  let start, finish;

  start = performance.now();
  for (let i = 0; i < 1000; i++) {
    await old.test.findUnique({
      select: {
        id: true,
        name: true,
      },
      where: {
        id: 1,
      },
    });
  }
  finish = performance.now();
  console.log(`2.14 took ${finish - start}ms`);

  start = performance.now();
  for (let i = 0; i < 1000; i++) {
    await prisma.test.findUnique({
      select: {
        id: true,
        name: true,
      },
      where: {
        id: 1,
      },
    });
  }
  finish = performance.now();
  console.log(`2.23 took ${finish - start}ms`);

  start = performance.now();
  for (let i = 0; i < 1000; i++) {
    await napi.test.findUnique({
      select: {
        id: true,
        name: true,
      },
      where: {
        id: 1,
      },
    });
  }
  finish = performance.now();
  console.log(`napi 2.23 took ${finish - start}ms`);
  prisma.$disconnect();
  old.$disconnect();
  napi.$disconnect();
}

main();
```


Here are the result that I am getting on Windows(AMD Ryzen 3900XT):
```
code/reproductions/issue-5792 is 📦 v1.0.0 via  v14.17.0
❯ yarn start
yarn run v1.22.5
$ ts-node main.ts
2.14 took 1496.481899999082ms
2.23 took 1499.4336000010371ms
napi 2.23 took 1120.0531999990344ms
Done in 5.80s.

code/reproductions/issue-5792 is 📦 v1.0.0 via  v14.17.0
❯ yarn start
yarn run v1.22.5
$ ts-node main.ts
2.14 took 1463.7449000030756ms
2.23 took 1497.6568999961019ms
napi 2.23 took 1122.1065999940038ms
Done in 5.74s.

code/reproductions/issue-5792 is 📦 v1.0.0 via  v14.17.0
❯ yarn start
yarn run v1.22.5
$ ts-node main.ts
2.14 took 1498.7672000005841ms
2.23 took 1469.091099999845ms
napi 2.23 took 1111.8403999954462ms
Done in 5.75s.

code/reproductions/issue-5792 is 📦 v1.0.0 via  v14.17.0
❯ yarn start
yarn run v1.22.5
$ ts-node main.ts
2.14 took 1478.4999000057578ms
2.23 took 1499.8465000018477ms
napi 2.23 took 1130.8392999991775ms
Done in 5.78s.
```

These result does not indicate a performance regression and aligns with what I expect. I will try this on MacOS now.

**Comment by pantharshit00** [comment]
I tried this on MacOS as well but I can't reproduce(2018 13 inch Macbook Pro):
```

╭─     ~/code/reproductions/issue_5792                                                ✔  took 4s    14.15.1    2.01G   ─╮
╰─ NODE_ENV=production node main.js                                                                                                 ─╯
2.14 took 2270.8000319600105ms
2.23 took 1108.5681910514832ms
napi 2.23 took 559.3272659778595ms

╭─     ~/code/reproductions/issue_5792                                                             ✔  14.15.1    1.98G   ─╮
╰─ NODE_ENV=production node main.js                                                                                                 ─╯
2.14 took 1011.5077039599419ms
2.23 took 1093.1392549872398ms
napi 2.23 took 549.714607000351ms
```

**Comment by Akxe** [comment]
@pantharshit00 please try my simple example with data provided to you under the NDA we have. The query is orders of magnitude slower since.

**Comment by k-sae** [comment]
I had the same issue while I was migrating from `sequelize`:

Environment:
```
os: Linux (5.10.42-1-MANJARO)
DB: MySQL (10.4.19-MariaDB)
Node: v16.3.0
prisma: 2.25
```
**Sequalize**:
```javascript
for (let index = 0; index < 2000; index++) {
    const allEmployees = await models.employees.findAll({
    limit: 100,
})
}
```
**Prisma**:
```javascript
for (let index = 0; index < 2000; index++) {
    const employees = await prisma.employees.findMany({
      take: 100
    })
}
```

Results:
```
prisma: 42.497s
sequelize: 8.709s
```

I don't know if the following issue #4911 is related however after enabling the `nodeApi` preview feature the performance greatly improved

Results with `nApi` enabled:
```
prisma: 6.146s
```

**Comment by janpio** [comment]
The Node-API preview feature does not change anything under the hood how we run queries, so this is probably a different thing than what was reported here before. Still great results you are seeing @k-sae. Maybe post these results in https://github.com/prisma/prisma/issues/6301 - and if you could even provide a reproduction we can try this out with that would be amazing.

**Comment by Akxe** [comment]
@janpio Can confirm, no change in the duration of my (prisma) query.

**Comment by k-sae** [comment]
@Akxe Have you run `npx prisma generate` after adding `previewFeatures = ["nApi"]` to schema.prisma?

**Comment by Akxe** [comment]
> @Akxe Have you run `npx prisma generate` after adding `previewFeatures = ["nApi"]` to schema.prisma?

Indeed I did, yet no change. Still it resulted in 36-38 sec

**Comment by k-sae** [comment]
@Akxe that is bad, the last thing I want is having performance issues on the server due to unknown parameters. 
I have conducted the test multiple times with logging disabled and the results are pretty close. 
Here some added information maybe it regards identifying the root cause of the issue.
The database I am using is the employee database from [here](https://dev.mysql.com/doc/employee/en/). It contains about 300,024 employee record 
**Prisma schema**: 
```
model employees {
  emp_no       Int              @id
  birth_date   DateTime         @db.Date
  first_name   String           @db.VarChar(14)
  last_name    String           @db.VarChar(16)
  gender       employees_gender
  hire_date    DateTime         @db.Date
}
```

The query generated from prisma: 
```SQL
SELECT `employees`.`employees`.`emp_no`, `employees`.`employees`.`birth_date`, `employees`.`employees`.`first_name`, `employees`.`employees`.`last_name`, `employees`.`employees`.`gender`, `employees`.`employees`.`hire_date` FROM `employees`.`employees` WHERE 1=1 ORDER BY `employees`.`employees`.`emp_no` ASC LIMIT ? OFFSET ?
```

The query generated from sequelize: 
```SQL
SELECT `emp_no`, `birth_date`, `first_name`, `last_name`, `gender`, `hire_date` FROM `employees` AS `employees` LIMIT 100;
```

I have run the both of the queries directly from the MySQL database and the result wasn't significantly large

Results
```
prisma raw query: 1.2ms
sequelize raw query: 0.9ms
```

The next step was the to test the sequelize-query using prisma `query.$raw`:
```javascript
 for (let index = 0; index < 2000; index++) {
  const allEmployeesRaw = await prisma.$queryRaw('SELECT `emp_no`, `birth_date`, `first_name`, `last_name`, `gender`, `hire_date` FROM `employees` AS `employees` LIMIT 100;');
 }
```

Results
```
prisma raw without nodeApi: 26.817s
prisma raw with nodeApi: 4.452s
```
Notice that in the case of raw prisma queries the performance improvement wasn't significant especially that I am now ignoring most of the prisma features.

**Comment by Akxe** [comment]
@k-sae Just to clarify, I am not from Prisma. In my case the SQL Prisma creates is not the bottleneck.

I ran my test. I did enable full SQL logging, and requested 

```javascript
prisma.carrierOfOffer.findMany({
select: {
  carrierID: true,
  shortID: true,
  type: true,
  periods: {
    select: {
      carrierID: true,
      groups: { select: { groupID: true, motiveID: true } }
    }
  },
  carrier: {
    select: {
      carrierID: true,
      regionID: true,
      districtID: true,
      lat: true,
      lng: true
    }
  },
  offerID: true,
}});
```

The SQL it produces is extremely long (and was posted previously), but when I copied the SQL and put it into MySQL workbench, it returned results under 0.1s. For me, the problem is within the processing of incoming result(s).

**Comment by janpio** [comment]
@k-sae As said above, your original problem was not the same as what this thread is about. You were experiencing a different problem, which is somehow solved by switching to Node-API - which will _not_ help for the problem @Akxe and Co describe here. No need to investigate further.

**Comment by k-sae** [comment]
@janpio Thank you, I will create a reproduction here #6301, maybe it will identifying this issue. I really appreciate the hard work of the Prisma team.

**Comment by peterrogov** [comment]
I am experiencing close to 10x difference in performance between operations executed by prisma and similar raw queries. I will make a reproduction repo soon. Just adding my voice for now. Prisma 3.8.0 and postgres 12 (macos)

**Comment by millsp** [comment]
Hey everyone, in Prisma 5 we have made considerable performance improvements. @GCastilho @peterrogov @k-sae @Akxe @chinanderm @Tsvetomir-M, if you are using Prisma 5 are you still experiencing these problems? Otherwise, I'd love if you could give it a try.

**Comment by millsp** [comment]
Hey everyone, given the age of this issue, the recent improvements in query and startup performance, and also the difficulty to reproduce, I am closing it. But don't take me wrong, if you are experiencing any performance regression, please open a new issue with the necessary information for us to reproduce it.
